lib/gs-app.c

```diff
@@ -3004,10 +3041,10 @@ gs_app_set_metadata (GsApp *app, const gchar *key, const gchar *value)
 	}
 	g_hash_table_insert (priv->metadata, g_strdup (key), g_strdup (value));
 
-	g_signal_emit (app,
-	               signals[SIGNAL_METADATA_CHANGED],
-	               g_quark_from_string (key),
-	               key);
+ emit_and_quit:
+	/* emit the signal from the mainloop */
+	helper = gs_metadata_signal_data_new (app, key);
```

Comment:
While I like this idea, it did not look like we actually hit any issues with the previous approach or assuming that metadata updates happen in the main loop. I am OK with the change if you think it's a good idea though.